internal/internal_worker_usage_collector.go

```diff
@@ -0,0 +1,126 @@
+package internal
+
+import (
+	"context"
+	"github.com/shirou/gopsutil/cpu"
+	"github.com/uber-go/tally"
+	"go.uber.org/cadence/internal/common/metrics"
+	"go.uber.org/zap"
+	"runtime"
+	"sync"
+	"time"
+)
+
+type (
+	workerUsageCollector struct {
+		workerType   string
+		cooldownTime time.Duration
+		logger       *zap.Logger
+		ctx          context.Context
+		wg           *sync.WaitGroup // graceful stop
+		cancel       context.CancelFunc
+		metricsScope tally.Scope
+	}
+
+	workerUsageCollectorOptions struct {
+		Enabled      bool
+		Cooldown     time.Duration
+		MetricsScope tally.Scope
+		WorkerType   string
+	}
+
+	hardwareUsage struct {
+		NumCPUCores     int
+		CPUPercent      float64
+		NumGoRoutines   int
+		TotalMemory     float64
+		MemoryUsedHeap  float64
+		MemoryUsedStack float64
+	}
+)
+
+func newWorkerUsageCollector(
+	options workerUsageCollectorOptions,
+	logger *zap.Logger,
+) *workerUsageCollector {
+	if !options.Enabled {
+		return nil
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	return &workerUsageCollector{
+		workerType:   options.WorkerType,
+		cooldownTime: options.Cooldown,
+		metricsScope: options.MetricsScope,
+		logger:       logger,
+		ctx:          ctx,
+		cancel:       cancel,
+		wg:           &sync.WaitGroup{},
+	}
+}
+
+func (w *workerUsageCollector) Start() {
+	w.wg.Add(1)
+	go func() {
+		defer func() {
+			if p := recover(); p != nil {
+				w.logger.Error("Unhandled panic in workerUsageCollector.")
+				w.logger.Error(p.(error).Error())
+			}
+		}()
+		defer w.wg.Done()
+		ticker := time.NewTicker(w.cooldownTime)
+		for {
+			select {
+			case <-w.ctx.Done():
+				return
+			case <-ticker.C:
+				// Given that decision worker and activity worker are running in the same host, we only need to collect
+				// hardware usage from one of them.
+				if w.workerType == "DecisionWorker" {
```

Comment:
this might not be future proof and also if customer is running separate processes for decision and activity workers then we will not have the hardware usage of those hosts that only runs activity workers. we should also not create no-op `workerUsageCollector`s if only one of them will do the work.
@Groxx what would be your recommendation for host level metric reporting on the client side? I would like to avoid global static variables but this use case probably requires one.

Reply:
We tried Sync.Once before, but that would cause issues with unit testing as it will just wait indefinitely for this routine to stop while blocking all other goroutine from closing

Reply:
you can override it in the unit tests
```
type once interface {
   Do(func())
}

var collectHardwareUsageOnce once
```

in typical startup this would be set to `sync.Once`:
```
collectHardwareUsageOnce = sync.Once{}
```

in test code you can initialize this to a fake implementation
```
collectHardwareUsageOnce = myFakeOnce{}  // myFakeOnce implements Do(func())
```

Reply:
Thank you for your suggestion. I have implemented that in the latest commit

Reply:
I don't see `EmitOnce` being used in workerUsageCollector. We should only have one (singleton) instance of `workerUsageCollector` which would be lazily created by the first worker instance. Rest of the workers would create a noOpUsageCollector. This lazy initialization logic should be hidden from workers. Worker just calls `newWorkerUsageCollector()` and that function should determine whether it's first time or not. Let's discuss offline  if more clarification needed.

Reply:
In our usecase, only the hardware info are once per host collected. Other worker type (decision worker and activity worker) should have different workerUsageCollector as they track different task type behaviors.

Reply:
what type of information are you planning to collect per worker basis in this `workerUsageCollector`?

Reply:
Tasklist backlog/poll response since decision and activity worker have their own pollers and that need to be scaled independently